File: dialects.py
#!/usr/bin/env python3
"""Read dialects.json, and turn one recorded prompt into the prompt a dialect actually sends.

The Python half of the dialect seam. `dialects.ts` is the other half and they read the SAME file,
for the reason providers.json states about itself: one hand-copied registry in this repository is
already one too many.

    python3 dialects.py                 # what is registered, and how many rules each dialect has
    python3 dialects.py --residuals     # every recorded prompt, transformed, and what it still owes

WHAT A DIALECT IS. A named, deterministic, TOTAL function from the prompt on record for an asset to
the prompt a provider set is sent. `literal` is the identity function and has zero rules. `positive`
restates every prohibition in the estate's briefs as an assertion about what IS in the frame.

WHY IT DOES NOT WEAKEN PARITY. Because the function is pure and its input is the reference set's own
record, a candidate in any dialect is provably answering a question DERIVED FROM the same asset's
recorded brief. verify.py --parity re-derives it and compares bytes, which is a stronger check than
the equality it used to run: equality could only say "these two strings differ", and this says
"this string is not what this dialect produces from the record".

    python3 -c "import dialects, json; print(dialects.apply('positive', json.load(open('MANIFEST.json'))['assets'][0]['prompt']))"
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from dataclasses import dataclass
from pathlib import Path

HERE = Path(__file__).resolve().parent
REGISTRY = HERE / "dialects.json"

#: `$1` in dialects.json becomes `\1` here. The ONE adapter in the seam, kept to a single line and
#: asserted from the TypeScript side (parity.test.ts) against this file's own output, so the two
#: loaders cannot come to disagree about what a rule means.
_BACKREF = re.compile(r"\$(\d)")


@dataclass(frozen=True)
class Rule:
    find: str
    replace: str
    regex: bool
    why: str

    def apply(self, text: str) -> str:
        if self.regex:
            return re.sub(self.find, _BACKREF.sub(r"\\\1", self.replace), text)
        return text.replace(self.find, self.replace)


@dataclass(frozen=True)
class Dialect:
    id: str
    label: str
    #: The dialect this one is derived FROM. None on the literal dialect, which is the record.
    source: str | None
    #: Whether a prompt in this dialect may still carry the negation vocabulary. False for the
    #: literal dialect, which is the estate's own prohibition-heavy style and is meant to.
    check_residuals: bool
    rules: tuple[Rule, ...]
    summary: tuple[str, ...]

    def apply(self, prompt: str) -> str:
        for rule in self.rules:
            prompt = rule.apply(prompt)
        return prompt


def _document() -> dict:
    return json.loads(REGISTRY.read_text())


def negation_vocabulary() -> tuple[str, ...]:
    return tuple(_document()["negationVocabulary"])


def load() -> list[Dialect]:
    document = _document()
    out = []
    for raw in document["dialects"]:
        out.append(
            Dialect(
                id=raw["id"],
                label=raw["label"],
                source=raw["source"],
                check_residuals=bool(raw["checkResiduals"]),
                summary=tuple(raw.get("summary", [])),
                rules=tuple(
                    Rule(
                        find=r["find"],
                        replace=r["replace"],
                        regex=bool(r.get("regex", False)),
                        why=r["why"],
                    )
                    for r in raw["rules"]
                ),
            )
        )
    return out


def by_id(dialect_id: str) -> Dialect:
    for dialect in load():
        if dialect.id == dialect_id:
            return dialect
    known = ", ".join(d.id for d in load())
    raise SystemExit(f"unknown dialect {dialect_id!r}; dialects.json registers: {known}")


def literal() -> Dialect:
    """The dialect the record is IN, and the one the controlled comparison is conducted in."""
    return by_id(_document()["literal"])


def apply(dialect_id: str, prompt: str) -> str:
    return by_id(dialect_id).apply(prompt)


def residuals(dialect_id: str, prompt: str) -> list[str]:
    """Negation vocabulary still present after the transform, lowercased and deduplicated.

    This is what makes "positive" a measured property rather than a claim. A prompt that still owes
    a residual may not be SENT in this dialect — promptForProvider throws, and this function is what
    verify.py reports the same fact from. The rule list therefore grows with the corpus instead of
    being written once and hoped over.
    """
    dialect = by_id(dialect_id)
    if not dialect.check_residuals:
        return []
    found: list[str] = []
    for word in negation_vocabulary():
        if re.search(rf"\b{re.escape(word)}\b", prompt, re.IGNORECASE) and word not in found:
            found.append(word)
    return found


def _residual_report() -> int:
    """Every recorded prompt in this repository, run through every checked dialect."""
    import providers  # local: only this entry point needs it, and providers.py does not need us

    reference = providers.reference()
    if not reference.exists:
        print("the reference set has no manifest; nothing to transform")
        return 1
    document = json.loads(reference.manifest.read_text())
    failures = 0
    for dialect in load():
        if not dialect.check_residuals:
            continue
        clean = 0
        for asset in document["assets"]:
            if asset.get("derivedFrom"):
                continue
            owed = residuals(dialect.id, dialect.apply(asset["prompt"]))
            if owed:
                failures += 1
                print(f"  {dialect.id}: {providers.key_of(asset)} still owes: {', '.join(owed)}")
            else:
                clean += 1
        print(f"{dialect.id}: {clean} recorded prompt(s) transform clean, {failures} still owe")
    return 1 if failures else 0


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__.split("\n", 1)[0])
    parser.add_argument("--residuals", action="store_true", help="transform every recorded prompt")
    args = parser.parse_args()
    if args.residuals:
        sys.exit(_residual_report())
    for entry in load():
        source = entry.source or "the record itself"
        print(f"{entry.id:<12} {len(entry.rules):>3} rule(s)  from {source:<18} {entry.label}")
    sys.exit(0)

File: dialects.ts
/**
 * The TypeScript half of the dialect seam. Reads the same `dialects.json` `dialects.py` reads.
 *
 * ══════════════════════════════════════════════════════════════════════════════════════════════
 * ## WHAT A DIALECT IS
 *
 * A named, deterministic, **total** function from the prompt on record for an asset to the prompt a
 * provider set is actually sent. `literal` is the identity function and carries zero rules;
 * `positive` restates every prohibition in this estate's briefs as an assertion about what IS in
 * the frame.
 *
 * ## WHY IT DOES NOT WEAKEN PROMPT PARITY
 *
 * `prompts.ts` guarantees that all providers replay the recorded prompt byte for byte. That
 * guarantee is what makes the FLUX-vs-Qwen comparison mean anything, and it is unchanged — it is
 * now scoped to a dialect and enforced three ways rather than one:
 *
 * 1. **Within a dialect, nothing has changed.** Every provider in a dialect is sent the identical
 *    string for a given asset. `parity.test.ts` asserts it per dialect group.
 *
 * 2. **Across dialects the check is STRONGER than the equality it replaces.** A dialect is a pure
 *    function whose input is the reference set's own record, so a candidate's prompt is not merely
 *    "different" — it is re-derivable. `verify.py --parity` applies the dialect's rules to the
 *    reference's recorded prompt and fails if one byte differs. Equality could only ever say "these
 *    two strings differ"; this says "this string is not what this dialect produces from the record".
 *
 * 3. **A set cannot be in a dialect by accident.** The dialect is on the provider entry, on every
 *    manifest entry, and in `compare.py`'s header, and `compare.py` refuses to print a
 *    single-question verdict over a selection that spans two of them. There is no path by which two
 *    sets generated from different prompts get compared without that being stated on the page.
 *
 * ## THE RESIDUAL RULE
 *
 * What makes "positive" a measured property rather than a claim. Each dialect declares the
 * vocabulary it forbids itself; after the rules run, the result is scanned for it, and a prompt
 * still carrying any of it **cannot be sent** — `promptForProvider` throws. An asset only enters the
 * positive set if the restatement genuinely cleared it, so the rule list grows with the corpus
 * rather than being written once and hoped over.
 * ══════════════════════════════════════════════════════════════════════════════════════════════
 */

import { readFileSync } from 'node:fs'
import { join } from 'node:path'

const HERE = import.meta.dirname

export interface DialectRule {
  readonly find: string
  readonly replace: string
  /** `$1`-style backreferences are permitted in `replace`. dialects.py translates them to `\1`. */
  readonly regex?: boolean
  readonly why: string
}

export interface Dialect {
  readonly id: string
  readonly label: string
  /** The dialect this one derives FROM. Null on the literal dialect, which is the record. */
  readonly source: string | null
  /**
   * Whether a prompt in this dialect may still carry the negation vocabulary. False for `literal`,
   * which is the estate's own prohibition-heavy house style and is supposed to.
   */
  readonly checkResiduals: boolean
  readonly summary: readonly string[]
  readonly rules: readonly DialectRule[]
}

interface DialectDocument {
  readonly literal: string
  readonly negationVocabulary: readonly string[]
  readonly dialects: readonly Dialect[]
}

const document = JSON.parse(readFileSync(join(HERE, 'dialects.json'), 'utf8')) as DialectDocument

export const DIALECTS: readonly Dialect[] = document.dialects
export const NEGATION_VOCABULARY: readonly string[] = document.negationVocabulary

export function dialectById(id: string): Dialect {
  const found = DIALECTS.find((d) => d.id === id)
  if (!found) {
    throw new Error(
      `unknown dialect "${id}"; dialects.json registers: ${DIALECTS.map((d) => d.id).join(', ')}`,
    )
  }
  return found
}

/** The dialect the record is IN, and the one the controlled comparison is conducted in. */
export const LITERAL: Dialect = dialectById(document.literal)

/** Apply one dialect's rules, in order, replacing every occurrence. Pure; no I/O; no provider. */
export function applyDialect(id: string, prompt: string): string {
  let out = prompt
  for (const rule of dialectById(id).rules) {
    out = rule.regex
      ? out.replace(new RegExp(rule.find, 'g'), rule.replace)
      : out.split(rule.find).join(rule.replace)
  }
  return out
}

/**
 * The negation vocabulary a transformed prompt still carries, deduplicated and in registry order.
 *
 * Empty for the literal dialect by construction: the estate's own style is prohibition-heavy, that
 * is the very thing under test, and flagging it would be flagging the control.
 */
export function residualNegations(id: string, prompt: string): readonly string[] {
  if (!dialectById(id).checkResiduals) return []
  const owed: string[] = []
  for (const word of NEGATION_VOCABULARY) {
    if (new RegExp(`\\b${word}\\b`, 'i').test(prompt) && !owed.includes(word)) owed.push(word)
  }
  return owed
}

/**
 * Thrown when a dialect's rules did not clear a prompt of its prohibitions.
 *
 * This is a refusal to SPEND, not a lint. The whole hypothesis being tested is "does Qwen honour a
 * brief with no prohibitions in it"; generating against a brief that still holds four of them would
 * answer a question nobody asked, on a deployment billed by the hour, and the resulting set would
 * carry a `dialect: positive` label that is not true of its own prompts.
 */
export class ResidualNegationError extends Error {
  constructor(dialect: string, key: string, owed: readonly string[]) {
    super(
      `${key}: the ${dialect} dialect left ${owed.length} prohibition word(s) in the prompt — ` +
        `${owed.join(', ')}. A set labelled "${dialect}" whose prompts still carry prohibitions ` +
        'would answer a different question from the one it claims to, so this asset cannot be ' +
        `generated until dialects.json has a rule that restates the clause holding it. Run ` +
        '`python3 dialects.py --residuals` to see every asset in this state.',
    )
    this.name = 'ResidualNegationError'
  }
}
